compiler/rustc_lint/src/gpukernel_abi.rs

```diff
@@ -0,0 +1,177 @@
+    /// ```
+    ///
+    /// ### Explanation
+    ///
+    /// `gpu-kernel` functions are usually searched by name in the compiled file.
+    /// A mangled name is usually unintentional as it would need to be searched by the mangled name.
+    ///
+    /// To use an unmangled name for the kernel, either `no_mangle` or `export_name` can be used.
+    /// ```rust,ignore (fails on non-GPU targets)
+    /// // Can be found by the name "kernel"
+    /// #[unsafe(no_mangle)]
+    /// extern "gpu-kernel" fn kernel() { }
+    ///
+    /// // Can be found by the name "new_name"
+    /// #[unsafe(export_name = "new_name")]
+    /// extern "gpu-kernel" fn other_kernel() { }
+    /// ```
+    MISSING_GPU_KERNEL_EXPORT_NAME,
+    Warn,
+    "mangled gpu-kernel function"
+}
+
+declare_lint_pass!(ImproperGpuKernelLint => [
+    IMPROPER_GPU_KERNEL_ARG,
+    MISSING_GPU_KERNEL_EXPORT_NAME,
+]);
+
+/// `ImproperGpuKernelLint` checks `gpu-kernel` function definitions:
+///
+/// - `extern "gpu-kernel" fn` arguments should be primitive types.
+/// - `extern "gpu-kernel" fn` should have an unmangled name.
+impl<'tcx> LateLintPass<'tcx> for ImproperGpuKernelLint {
+    fn check_fn(
+        &mut self,
+        cx: &LateContext<'tcx>,
+        kind: hir::intravisit::FnKind<'tcx>,
+        decl: &'tcx hir::FnDecl<'_>,
+        _: &'tcx hir::Body<'_>,
+        span: Span,
+        id: LocalDefId,
+    ) {
+        use hir::intravisit::FnKind;
+
+        let abi = match kind {
+            FnKind::ItemFn(_, _, header, ..) => header.abi,
+            FnKind::Method(_, sig, ..) => sig.header.abi,
+            _ => return,
+        };
+
+        if abi != ExternAbi::GpuKernel {
+            return;
+        }
+
+        let sig = cx.tcx.fn_sig(id).instantiate_identity();
+        let sig = cx.tcx.instantiate_bound_regions_with_erased(sig);
+
+        for (input_ty, input_hir) in iter::zip(sig.inputs(), decl.inputs) {
+            let is_valid_arg = match input_ty.kind() {
+                ty::Bool
+                | ty::Char
+                | ty::Int(_)
+                | ty::Uint(_)
+                | ty::Float(_)
+                | ty::RawPtr(_, _) => true,
+
+                ty::Adt(_, _)
+                | ty::Alias(_, _)
+                | ty::Array(_, _)
+                | ty::Bound(_, _)
+                | ty::Closure(_, _)
+                | ty::Coroutine(_, _)
+                | ty::CoroutineClosure(_, _)
+                | ty::CoroutineWitness(..)
+                | ty::Dynamic(_, _)
+                | ty::Error(_)
+                | ty::FnDef(_, _)
+                | ty::FnPtr(..)
+                | ty::Foreign(_)
+                | ty::Infer(_)
+                | ty::Never
+                | ty::Param(_)
+                | ty::Pat(_, _)
+                | ty::Placeholder(_)
+                | ty::Ref(_, _, _)
+                | ty::Slice(_)
+                | ty::Str
+                | ty::Tuple(_)
+                | ty::UnsafeBinder(_) => false,
```

Comment:
The answer for `ty::UnsafeBinder` should be the recursive answer to this question by default (I think the implied result is quite useless given the answer to other things, but that's fine).